apps/devhub/tasks.py

```diff
@@ -32,65 +32,116 @@
-        raise
+        # FIXME: Have validator return native Python datastructure.
+        return json.loads(result)
 
-    try:
-        # Try annotating the result with the "passed auto validation" status.
-        validation = json.loads(result)
-        # The automatic validation (which will automatically review and sign
-        # the file) is for unlisted (non-sideload) addons and beta versions.
-        # It only passes this automatic validation if there's no message with a
-        # signing_severity above trivial (so if there's any low/medium/high
-        # severity, it fails).
-        # Any "regular" messages from the amo-validator that should fail the
-        # auto validation now have a signing_severity level attached.
-        passed_auto_validation = False
-        if 'signing_summary' in validation:
-            summary = validation['signing_summary']
-            passed_auto_validation = (summary['low'] == 0 and
-                                      summary['medium'] == 0 and
-                                      summary['high'] == 0)
-        validation['passed_auto_validation'] = passed_auto_validation
-        result = json.dumps(validation)
-    except Exception:  # A bare `except:` would ignore system exceptions.
-        # Failed loading the result? No need to set passed_auto_validation.
-        pass
-
-    # Skip the "addon signed" warning if we're not signing.
-    if not settings.SIGNING_SERVER:
-        result = skip_signing_warning(result)
-    upload.validation = result
+
+@task(soft_time_limit=settings.VALIDATOR_TIMEOUT)
+def validate_file(file_id, **kw):
+    """Validate a File instance. If cached validation results exist, return
+    those, otherwise run the validator."""
+
+    if settings.VALIDATE_ADDONS:
+        file = File.objects.get(pk=file_id)
+        try:
+            return file.validation.validation
+        except FileValidation.DoesNotExist:
+            result = run_validator(file.current_file_path,
+                                   listed=file.version.addon.is_listed)
+            # FIXME: Have validator return native Python datastructure.
+            return json.loads(result)
+
+
+@task
+@write
+def handle_upload_validation_result(results, upload_id):
+    """Annotates a set of validation results and saves them to the given
+    FileUpload instance."""
+    result = annotate_validation_results(results)
+
+    upload = FileUpload.objects.get(pk=upload_id)
+    upload.validation = json.dumps(result)
     upload.save()  # We want to hit the custom save().
+    return upload
 
 
-def skip_signing_warning(result_json):
+@task
+@write
+def handle_file_validation_result(results, file_id):
+    """Annotates a set of validation results and saves them to the given
+    File instance."""
+    result = annotate_validation_results(results)
+
+    file_ = File.objects.get(pk=file_id)
+    return FileValidation.from_json(file_, result)
+
+
+@task
+def annotate_validation_results(results):
```

Comment:
I'm not a big fan of functions that have two different behaviors depending on the type of parameters they get (eg they do one thing if they receive a dict, or another one if they receive a list).

Could this be split?